### ISSUE TITLE
New Installation problems 1.4.0+: please update documentation

### ISSUE THREAD

**Issue by otheus** [alextselegidis/easyappointments#1077]
Only after reviewing all relevant (and closed) tickets, was I able to make a fresh EA installation using v1.4.1.  Please update the documentation for both v1.4.0 and v1.4.1.

## Problem 1: missing vendor/autoload.php
Server generates a 500 error, and the PHP-error log shows  something like 
```
require_once(..../vendor/autoload.php): failed to open stream
``` 
Solution is to run composer.
```
  composer upgrade
```
There is a composer.lock file in the published v1.4.1 tag, and this is IMO probably incorrect. Either it should be shipped with vendor libs already installed, or no composer.lock. 

I'm not sure if npm is needed or if that is system specific. 

Referenced also by:
 * https://github.com/alextselegidis/easyappointments/issues/933
 * https://github.com/alextselegidis/easyappointments/issues/338

## Problem 2: Session data not writable. 
The documentation suggests that the `storage` should be writable, but does not mention sub-directories.  These must be writable as well. The suggested command to ensure these are writable is:
```
  find storage -type d -exec chmod o+rwx {} \;
```
Of course, if someone installs the software as non-root, as one should, then it makes sense to set the group ownership of all files to, for instance, `apache` (or the nginx user), and use:
```
  find storage -type d -exec chmod 2770 {} \;
```
Referenced also in: https://github.com/alextselegidis/easyappointments/issues/858

## Problem 3: Wrong installation URL
The current installation documents say one should go to `http://installation-url/index.php/backend`. That was true in v1.3.2, but since 1.4.0, this redirects to a login and fails with a blank page, or with debugging on, a notice indicates the program is looking for tables that it has yet to install. This is probably not a problem with existing installations. (untested)
```
Error Number: 1146

Table 'ea_iezw.ea_settings' doesn't exist

SELECT * FROM `ea_settings` WHERE `name` = 'company_name'

Filename: models/Settings_model.php

Line Number: 40
```

The solution is suggested in several issues, and is to go to the URL `http://localhost:81/easyappointments/index.php/installation/index`

References:
 * https://github.com/alextselegidis/easyappointments/issues/605
 * https://github.com/alextselegidis/easyappointments/issues/979
 * https://github.com/alextselegidis/easyappointments/issues/886

## Problem 4: Missing *.min.css and *.min.js

Apparently, CSS and JS files are not found.

## Problem 5: npm start: wtf?

A few issues here mention using `npm install` and `npm start`.  Setting aside the fact these shouldn't be necessary, it's not clear if `npm start` needs to be run after every boot. It's also not clear if you can have multiple instances running in parallel on the same host, or if these will interfere with each other.

## Problem 6: npm start: npx not found / SyntaxError

`npm start` fails with `npx not found` on RHEL7 systems. This is because it uses npm v3.10. and npx comes prepackaged only starting sometime in 5.x. 

However, it's worse than that, as apparently, you have developed against a version that is too new.
```
 npm start

> ea-iezw@1.0.0 start /sites/...../ea   > npx gulp

/sites/.../ea/node_modules/del/index.js:55
module.exports = async (patterns, {force, dryRun, cwd = process.cwd(), ...options} = {}) => {
                                                                       ^^^

SyntaxError: Unexpected token ...
    at createScript (vm.js:56:10)
    at Object.runInThisContext (vm.js:97:10)
    at Module._compile (module.js:549:28)
    at Object.Module._extensions..js (module.js:586:10)
    at Module.load (module.js:494:32)
    at tryModuleLoad (module.js:453:12)
    at Function.Module._load (module.js:445:3)
    at Module.require (module.js:504:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/sites/..../ea/gulpfile.js:18:13)
```

Note: the real directory was replaced with `...` for security reasons.

**Comment by otheus** [comment]
Honestly, you had a great product, but you fell into the docker-lifecycle trap, which is basically a recursively, self-reinforcing dependency loop that forces everyone to use docker because that's the only way you can make software work with all the non-interchangeable dependencies, which entices developers to use only the most recent releases of these dependencies, which forces more dependency on docker, etc.

**Comment by alextselegidis** [comment]
Hello! 

Thanks for the detailed information on this ticket. 

The installation instructions of the website refer to the pre-built version you can get from the website and not from the repository. 

Problem 1: 

The README file of this repo states that you need to run composer install to get the PHP dependencies locally. 

Problem 2:

That was implied as this is how almost all modern PHP apps are working.

Problem 3: 

The URL is a placeholder, you will need to use your own `installation-url`.

Problem 4: 

For this repo I saw that the `npm run build` step is missing from the README file so that has to be added. 

Problem 5: 

The file watcher is necessary if you develop as it will automatically build your files while you change them. 

Problem 6: 

The npx utility is published around 4 years ago, please install the latest node LTS and you will have no problems.

<small>
<table>
<tr>
<td><img src="https://easyappointments.org/img/easyappointments-signature.png"></td>
<td>&nbsp;&nbsp;Alex Tselegidis, Easy!Appointments Creator
<br>&nbsp;&nbsp;<a href="https://easyappointments.org/get-a-free-quote">Need a customization? Get a free quote!</a></td>
</tr>
</table>
</small>

**Comment by noraj** [comment]
The docker should contain the build process including npm and composer, etc.

**Comment by otheus** [comment]
Hi @alextselegidis  --I return to this project after a year. Now I use docker and  alpine and php7.3 and all the latest gadgets, ie, npm, composer.  Now, npm, composer, npx work nicely. 

Problem #1: My problems originated from using git to customize stuff and then re-deploy. In 1.4, you added a .gitignore, which excluded vendor files. Thus, these were missing from re-deployment. 

now I'm trying to build things via Docker. 

Problem #2: Permissions  for storage. The documentation is still incomplete about this.

Problem #3: I think you misunderstood. Of course `installation-url` was  a placeholder. The problem was the rest of the URL.  You meant, I think `<.... /installation`

Problem #4: `npm run build` does not help -- it creates a directory called "build" and a zip file containing the things already in the directory. it does NOT create the missing `.min.js`  and `.min.css` files. Where are these coming from? 

For instance, after performing the `npm install && composer install`, the browser console sees dozens of 404s, for instance: `http://localhost/assets/ext/bootstrap/css/bootstrap.min.css` is missing.  They are also not in the `build/` tree after `npm run build`.

Have you tried a clean build with 1.4? Or 1.5?

**Comment by alextselegidis** [comment]
Hello! 

Thanks for posting this, a complete and detailed documentation is very important indeed, I will update the file accordingly. 

The installation guide does not apply to a git clone, it applies to the pre-build release package you can find under Github Releases and easyappointments.org. That must apparently become more clear. 

Additionally the minified files and the 3rd party dependencies are not part of the repo because they are dynamically generated during the development or build process. 

Hope that helps :)

<small>
<table>
<tr>
<td><img src="https://easyappointments.org/img/easyappointments-signature.png" style="margin-right: 5px"></td>
<td style="line-height: 2">Alex Tselegidis, Easy!Appointments Creator <br><a href="https://easyappointments.org/get-a-free-quote">Need a customization? Get a free quote!</a></td>
</tr>
</table>
</small>

**Comment by otheus** [comment]
Alex -- I did figure it out. My Dockerfile varies much from the others presented here. It depends on local customization. Still WIP.

Note: I prefer systems based on runit. The main reason is that you can restart the apache engine without having to kill the container. The second reason is you can flexibly add supporting services such as postfix, cron, mysql, etc. This is a multi-stage dev build based on Alpine. A Makefile is used to create some of the files that are COPY'd from the host OS. There is a notable lack of Environment variables. I'm not a big fan of using these, and they don't play well with runit -- not without some additional scaffolding. 

The resulting prod build -- without the DB  -- is 140 MB on our Linux x86_64 systems.


```
FROM alpine:3.10 AS base_php
# disable package prompt interaction

# install runit (init scheme)
RUN apk update && apk add \
	mariadb-client \
	php7 \
	php7-mysqli \
	php7-session \
	php7-xdebug \
	php7-opcache \
	php7-mysqli \
	php7-mbstring \
	php7-curl \
	php7-gd \
	php7-ctype \
	php7-dom \
	php7-json \
	php7-tokenizer \
	php7-xmlwriter \
	php7-xml \
	tzdata

#################################################################################################
### BUILD DEPLOYABLE APP                   ######################################################
#################################################################################################
FROM base_php AS prep_app

RUN adduser -D builder
RUN apk update && apk add npm composer 

USER builder
WORKDIR /sources
COPY --chown=builder sources/  /sources

# Prepare the building process
RUN   npm install && composer install 

# Build
RUN   npm run build \
      && cd /sources/build \
      && chmod -R a+rX,o-w . 

#################################################################################################
### PREPARE MYSQL SERVICE AND LOAD DATASET ######################################################
#################################################################################################
FROM alpine:3.10 AS prep_db

RUN apk update && apk add mariadb-client mariadb 
RUN apk add runit
RUN echo $RANDOM

ADD scripts/mysql-init.runit /etc/service/mysql/run
ADD data/db-init.sql /var/tmp/
RUN /bin/sh -x /etc/service/mysql/run

#################################################################################################
#### CREATE DEPLOYABLE STAGING CONTAINER ########################################################
#################################################################################################
FROM base_php
RUN apk update && apk add \
	runit \
	apache2 \
	apache2-utils \
	apache2-ctl \
	php7-apache2 

## DEVONLY RUN apk add mariadb \

STOPSIGNAL SIGCONT

COPY --from=prep_app --chown=root:root      /sources/build/  /var/www/localhost/htdocs/
COPY --from=prep_app --chown=apache:apache  /sources/build/storage/  /var/www/localhost/htdocs/storage/
## DEVONLY:  COPY --from=prep_db  --chown=mysql:mysql    /var/lib/mysql/  /var/lib/mysql/

COPY scripts/firstrun /usr/local/sbin/
# configure apache service by using runit directive
COPY scripts/apache.runit /etc/service/apache/run
# mysql-support-install-script
##  DEVONLY:  COPY scripts/mysql-init.runit /etc/service/mysql/run
# Apache config:
COPY configs/apache-docker.conf /etc/apache2/conf.d/docker.conf
COPY configs/apache-ea.conf     /etc/apache2/conf.d/ea.conf
COPY configs/php7.conf	        /etc/php7/conf.d/99_local.conf

# install and configure app stuff
## COPY data/ea-config.php /var/www/localhost/htdocs/config.php
COPY data/ea-config.php /var/www/localhost/htdocs/config.php
COPY configs/ea-app-config.php /var/www/localhost/htdocs/application/config/config.php
COPY configs/ea-app-email.php /var/www/localhost/htdocs/application/config/email.php

# Create runit service directories, log file dirs, 
# ensure permissions in volatile area is correct,
# move app/log dir to /var/log, etc
# install runit
RUN set -e ;\
    mkdir -p /etc/runit ;\
    find /etc/service -type f -exec chmod +x {} +  ;\
    \
    rm -f /etc/apache2/conf.d/userdir.conf ;\
    \
    test -d /var/log/php || \
      install -d -o apache -g root -m 0770 /var/log/php ;\
    \
    /bin/true test -d /var/log/mysql || \
      /bin/true install -d -o mysql -g mysql -m 0770 /var/log/mysql ;\
    \
    cd /var/www/localhost/htdocs/storage ;\
      mv logs /var/log/ea ;\
      chown -R apache.apache /var/log/ea ;\
      ln -snf /var/log/ea logs ;\
    \
    /bin/true

# ports and entrypoint configuration
EXPOSE 80
ENTRYPOINT exec runsvdir -P /etc/service 2>&1 'log: ................................................................................................................................' 
```
